[Core][FastPR] Fix OMP issue in particles utilities

pablobecker opened the pull request:
**📝 Description**
Fixes the ocassional failures in the particles utilities test. See https://stackoverflow.com/questions/39037239/c-openmp-and-stdvectorbool

**🆕 Changelog**
- Adds an aux int vector that stores if the elem was found or not
-  then the int vector is stored in the bool vector to be returned, to avoid changing the interface

pablobecker committed:
fix omp problem

matekelemen commented on kratos/utilities/particles_utilities.h:
```diff
@@ -252,17 +252,18 @@ class ParticlesUtilities
         typename BinBasedFastPointLocator<TDim>::ResultContainerType TLS(max_results);
 
         auto interpolations = std::make_pair(std::vector<bool>(rCoordinates.size1()), std::vector<TDataType>(rCoordinates.size1()));
+        std::vector<int> is_found_aux_vector(rCoordinates.size1()); //aux vector using ints to allow running in omp
```
```suggestion
        std::vector<char> is_found_aux_vector(rCoordinates.size1()); //aux vector using ints to allow running in omp
```

`int` is an overkill for this purpose

loumalouomega replied:
This is more efficient, but using char for not text is strange for me

RiccardoRossi replied:
we  could return DenseVector< bool>

matekelemen commented on kratos/utilities/particles_utilities.h:
```diff
@@ -280,6 +281,11 @@ class ParticlesUtilities
+        //storing info in bool vector (do not use omp!)
+        for(unsigned int i=0; i<rCoordinates.size1(); i++){
+            (interpolations.first)[i] = static_cast<bool>(is_found_aux_vector[i]);
+        }
```
```suggestion
        std::copy(is_found_aux_vector.begin(), is_found_aux_vector.end(), interpolations.first.begin());
```

matekelemen commented on kratos/utilities/particles_utilities.h:
```diff
@@ -252,17 +252,18 @@ class ParticlesUtilities
             const bool is_found = rLocator.FindPointOnMesh(row(rCoordinates,i), shape_functions, p_element, rTLS.begin(), rTLS.size(), SearchTolerance);
 
-            (interpolations.first)[i] = is_found;
+            is_found_aux_vector[i] = static_cast<int>(is_found);
```
```suggestion
            is_found_aux_vector[i] = static_cast<char>(is_found);
```

matekelemen reviewed (commented):
I've no idea how deep this function is used in the app, but if you have a bit of wiggle room with the return type, consider changing it to a vector of pairs instead of the current pair of vectors. It would eliminate the race condition and cut memory allocations in half.
```diff
- std::pair<std::vector<bool>, std::vector<T>>
+ std::vector<std::pair<bool,T>>
```

pablobecker committed:
@matekelemen review

RiccardoRossi commented on kratos/utilities/particles_utilities.h:
```diff
@@ -252,17 +252,18 @@ class ParticlesUtilities
         typename BinBasedFastPointLocator<TDim>::ResultContainerType TLS(max_results);
 
         auto interpolations = std::make_pair(std::vector<bool>(rCoordinates.size1()), std::vector<TDataType>(rCoordinates.size1()));
+        std::vector<char> is_found_aux_vector(rCoordinates.size1()); //aux vector using ints to allow running in omp
```
let's return DenseVector<bool>

this is a stupid mistake from my side. Sorry about this

loumalouomega committed:
Merge branch 'master' into core/fix-particles-utilities

loumalouomega committed:
Revert "@matekelemen review"

This reverts commit 0f4ee322b46824616675dd8963aa399d249ff10e.

loumalouomega committed:
Revert "fix omp problem"

This reverts commit f167d797b2ffc4aea57dfd6fffa787ebdf5bbf5f.

loumalouomega committed:
Clean up

loumalouomega committed:
Fix DenseVector

loumalouomega committed:
Separate InterpolateValuesAtCoordinates and ImplementationInterpolateValuesAtCoordinates

RiccardoRossi commented on kratos/utilities/particles_utilities.h:
```diff
@@ -276,26 +274,48 @@ class ParticlesUtilities
+           - values the interpolated values (only valid if the corresponding value of is_inside is true)
+    */
+    template<unsigned int TDim, class TDataType, bool InterpolationVariableHasHistory>
+    static std::pair<std::vector<bool>, std::vector<TDataType> > InterpolateValuesAtCoordinates(
```
here you are using again std::vector bool

loumalouomega replied:
It is because pybind, doesn't work if not using std::vector or any compatible class. In all cases copy is necessary

RiccardoRossi reviewed (changes requested):
this PR makes a copy which should be avoided!!

loumalouomega committed:
Expose DenseVectorBool

loumalouomega committed:
Revert implementation

matekelemen commented on kratos/python/add_vector_to_python.cpp:
```diff
@@ -455,6 +450,44 @@ void  AddVectorToPython(pybind11::module& m)
+    bool_vector_binder.def(py::init( [](py::buffer b)
+    {
+        py::buffer_info info = b.request();
+        KRATOS_ERROR_IF( info.format != py::format_descriptor<typename DenseVector<bool>::value_type >::value ) << "Expected a double array\n";
```
```suggestion
        KRATOS_ERROR_IF( info.format != py::format_descriptor<typename DenseVector<bool>::value_type >::value ) << "Expecting a bool vector\n";
```

matekelemen commented on kratos/python/add_vector_to_python.cpp:
```diff
@@ -455,6 +450,44 @@ void  AddVectorToPython(pybind11::module& m)
+
+        for( int i=0; i<info.shape[0]; ++i ) {
+            vec[i]= static_cast<typename DenseVector<bool>::value_type *>(info.ptr)[i];
+        }
```
```suggestion
        std::copy(
            static_cast<const bool*>(info.ptr),
            static_cast<const bool*>(info.ptr) + info.shape[0],
            vec.begin()
        );
```

matekelemen reviewed (commented):
I still recommend returning `std::vector<std::pair<bool,TDataType>>`.

loumalouomega committed:
Comment

Co-authored-by: Máté Kelemen <44344022+matekelemen@users.noreply.github.com>

loumalouomega committed:
Better

Co-authored-by: Máté Kelemen <44344022+matekelemen@users.noreply.github.com>

loumalouomega committed:
Minor

RiccardoRossi committed:
Update add_vector_to_python.cpp

limiting interface of BoolVector. It makes no sense to have multiplication and division for a vector of bools

loumalouomega commented on kratos/python/add_vector_to_python.cpp:
```diff
@@ -455,6 +451,50 @@ void  AddVectorToPython(pybind11::module& m)
 
+    //***********************************************************************************
+        //***********************************************************************************
+    py::class_< DenseVector<bool>, std::shared_ptr<DenseVector<bool>> > bool_vector_binder(m,"BoolVector", py::buffer_protocol());
```
I was going to add some constexpr to use the already existing definitions and minimize code duplication

RiccardoRossi replied:
well, the vector interface could be definitely be improved...however i would do it in a different Pr. What do you think?

loumalouomega reviewed (approved):
@RiccardoRossi you can remove now the block
